### PR TITLE
Feat: Improve failed auth handling

### LINKED ISSUE #3523: QA TX: Handling slow transactions results in duplicate transactions
## Steps to reproduce

1. Create a new team action.
2. Submit the action, and sign the tx in metamask.
3. The transaction is signed, the UI shows the pending pill in the action header as completed, and the blue pill hides.
4. The primary CTA in the action footer continues to show as pending for a number of seconds after. 
5. The UI does not show the action as complete, and the CTA returns back to the default state in the action and I can submit the transaction again.
6. I do not receive any userhub errors.
7. I can continue this same process and can sign the same transaction over and over again. 
8. Sometimes it will then go to a completed state.
9. Multiple transactions then show in the activity hub. 

Set up: using Colony's RPC endpoint in metamask on Mac OS in Firefox with adblockers disabled. 

## Expected behaviour

* Upon signing the transaction, the action should stay in a pending state until the transaction is complete and then it should redirect the user to the completed action panel. 

## Actual behaviour

Recording of behaviour:

https://github.com/user-attachments/assets/7ddf55de-44dc-48b7-92f4-79bcf10bae51

The result of signing multiple transactions without the action being completed. Note the title of the action has also been replaced. 

image

Another example, where one transaction shows the correct title and the second transaction title was replaced:
image

### PR DESCRIPTION
**Pull request by iamsamgibbs**
## Description

This PR attempts to improve the way we handle failed mutations. There are few instances on QA and Production where actions were failing due to apollo 403 errors, the transaction would still succeed but the form would not proceed to the completed action screen due to the failed apollo calls. This meant the form could be resubmitted over and over resulting in multiple transactions.

It is likely these issues are caused by auth proxy rejecting the mutation. This can happen if auth proxy restarted after the saga has started running as the user will no longer be authenticated.

This PR attempts to improve this experiencing by adding a retry with re-authentication mechanism to all mutations and authentication attempts. If auth proxy rejects the mutation it will prompt the user to sign in again with metamask and reattempt the mutation. If the mutation still fails after this, the form will show an error message and the submit button will be disabled. This is a real edge case scenario and hopefully will never actually be seen in production.

## Testing

This is going to be a bit of a pain to test I'm afraid and involves running a deliberately broken version of auth proxy locally to facilitate testing.

* Step 1 - Spin up the dev environment as usual and run the create-data script.
* Step 2 - Clear your metamask activity and nonce data by opening the metamask extension, going to settings, advanced and then clicking the "Clear activity tab data" button.
* Step 3 - Connect your metamask wallet and open the Planex colony.
* Step 4 - In docker, stop the auth container.
* Step 5 - We need to run the auth proxy locally, clone down the repository (https://github.com/JoinColony/colonyCDappAuthProxy) and run `npm install` if you have never used it before.
* Step 6 - Switch to the `tmp/3669-testing` branch on auth proxy. I've made some changes here which will cause the following calls to occasionally fail (this obviously won't be merged in and is just for the sake of testing the changes here): createTransaction, updateTransaction, createDomainMetadata. The authentication check and handler functions will also occasionally fail. The updateDomainMetadata call will always fail.
* Step 7 - Run `npm run dev` on the auth proxy.

From this point on, if your wallet is disconnected and you end up getting kicked out of the colony, it may take a couple of attempts to reconnect the wallet. This is due to the auth proxy changes. I've not added retries to the main wallet connect authentication as we haven't had issues raised in this area.

* Step 8 - Okay, back on the Planex colony! Create a new team action and submit the form. Due to the auth proxy changes, we are expecting some of the calls to fail. The CDapp changes introduced in this PR allow us to reattempt these calls and we will be prompted to sign in with metamask when this happens. Whenever a metamask pop up appears, accept the transactions and sign in prompts. Eventually you should make it through.

https://github.com/user-attachments/assets/38da06c1-8678-4862-a198-ccaa3b0ed6e1

Compared to master, there is no automatic retry mechanism so the submit button will simply get stuck on "Pending" when a call fails.

* Step 9 - Now let's try an "Edit team" action. Submit the form. Due to the auth proxy changes we are expecting this one to fail completely. Accept your way through the metamask prompts until it gets to the metadata mutation which should fail. An error message should appear and the submit button should be disabled to prevent the user from reattempting the transaction. (Note: If the transaction itself fails, the UserHub will open to show the transaction error and the form itself will not show the new apollo error message.)

https://github.com/user-attachments/assets/8c019c4a-641b-401a-8388-ce3e0ac1e954

Screenshot 2024-11-13 at 11 58 06

* Step 10 - Feel free to try out some other actions, they should all work similarly to the create team action.

### Further testing

* Step 11 - Kill your local auth proxy instance, and restart the auth container in docker.

* Step 12 - Okay we're back to a bit more of a normal environment now. Try running a bunch of actions and check nothing has broken. Test this with both the metamask wallet and the dev wallet.

## Diffs

**New stuff** ✨

* Added `mutateWithAuthRetry` function
* Added `authenticateWalletWithRetry` function

**Changes** 🏗

* Action form calls `authenticateWalletWithRetry` instead of `authenticateWallet` before calling saga
* Action form now shows an error on the frontend when there is an apollo error.
* All instances of `apolloClient.mutate` are now wrapped in `mutateWithAuthRetry`

## TODO

- [ ] Delete the `tmp/3669-testing` branch on auth proxy once this is merged
- [ ] Remove the transaction retry mechanism to come in a later PR

Resolves #3523
Contributes to #3510

### CODE REVIEW

**Review by bassgeta — APPROVED** [review]
This is such a nice API, just wrap it in `mutateWithAuthRetry` and that will trigger wallet reauth (also in a separate helper, very clean!) :muscle: 
Retries went through 2 times :heavy_check_mark: 

https://github.com/user-attachments/assets/30b7b97b-85f4-4f1e-8f07-bd8126f8cbfa

Edit team fails nicely :heavy_check_mark: 

https://github.com/user-attachments/assets/c8669dcd-302b-4d6a-bd3b-281a10bcaeb0

I have also tried rejecting a transaction and this error is shown (intentional or not :eyes: ) 
image

File: src/apollo/utils.ts [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import { authenticateWalletWithRetry } from '~auth';
+
+export async function mutateWithAuthRetry(
+  mutationFn: () => Promise<any>,
+  maxRetries = 3,
+) {
+  let attempts = 0;
+  let result: any;
+  while (attempts < maxRetries) {
```
Comment (olgapod) [inline src/apollo/utils.ts:9]: Hey @iamsamgibbs just a quick thought – since we already have `exponential-backoff` lib in our package.json, would it make sense to refactor this while loop with that library?

Comment (iamsamgibbs) [reply src/apollo/utils.ts:9]: Oh interesting, I didn't realise we had this package. I will take a look.

**Comment by iamsamgibbs** [comment]
Thanks for raising that @olgapod - I've refactored the functions to use exponential-backoff.

File: src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/ActionSidebarContent.tsx [hunk 2]
```diff
@@ -283,6 +303,29 @@ const ActionSidebarContent: FC<ActionSidebarContentProps> = ({
           key={actionFormProps.mode}
           className="flex h-full flex-col"
           innerRef={formRef}
+          onError={(error) => {
+            console.error(error);
+
+            // Auth failed before transaction was sent
+            if (
+              (error.message as unknown) === 'Auth failed' ||
+              (error.code as unknown) === 4001
```
[the remark this answers is not in the document]
Comment (iamsamgibbs) [reply src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/ActionSidebarContent.tsx:312]: Metamask throws an error with a 4001 error code when the user manually rejects the transaction :)

[the remark this answers is not in the document]
Comment (bassgeta) [reply src/components/v5/common/ActionSidebar/partials/ActionSidebarContent/ActionSidebarContent.tsx:312]: :saluting_face:

**Review by bassgeta — APPROVED** [review]
Exponentially reapproving :D 

retested creating a team, went through the 3rd time
image
edit fails :heavy_check_mark: 
image

**Review by rdig — APPROVED** [review]
